homeassistant/components/nanoleaf/config_flow.py

```diff
@@ -0,0 +1,221 @@
+                errors={"base": "cannot_connect"},
+                last_step=False,
+            )
+        except NotAuthorizingNewTokens:
+            pass
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unknown error connecting to Nanoleaf")
+            return self.async_show_form(
+                step_id="user",
+                data_schema=USER_SCHEMA,
+                last_step=False,
+                errors={"base": "unknown"},
+            )
+        return await self.async_step_link()
+
+    async def async_step_zeroconf(
+        self, discovery_info: DiscoveryInfoType
+    ) -> FlowResult:
+        """Handle Nanoleaf Zeroconf discovery."""
+        _LOGGER.debug("Zeroconf discovered: %s", discovery_info)
+        return await self._async_discovery_handler(discovery_info)
+
+    async def async_step_homekit(self, discovery_info: DiscoveryInfoType) -> FlowResult:
+        """Handle Nanoleaf Homekit discovery."""
+        _LOGGER.debug("Homekit discovered: %s", discovery_info)
+        return await self._async_discovery_handler(discovery_info)
+
+    async def _async_discovery_handler(
+        self, discovery_info: DiscoveryInfoType
+    ) -> FlowResult:
+        """Handle Nanoleaf discovery."""
+        host = discovery_info["host"]
+        # The name is unique and printed on the device and cannot be changed.
+        name = discovery_info["name"].replace(f".{discovery_info['type']}", "")
+        await self.async_set_unique_id(name)
+        self._abort_if_unique_id_configured({CONF_HOST: host})
+        self.nanoleaf = Nanoleaf(host)
+
+        # Import from discovery integration
+        self.device_id = discovery_info["properties"]["id"]
+        self.discovery_conf = cast(
+            dict,
+            await self.hass.async_add_executor_job(
+                load_json, self.hass.config.path(CONFIG_FILE)
+            ),
+        )
+        self.nanoleaf.token = self.discovery_conf.get(self.device_id, {}).get(
+            "token",  # >= 2021.4
+            self.discovery_conf.get(host, {}).get("token"),  # < 2021.4
+        )
+        if self.nanoleaf.token is not None:
+            self.context["source"] = config_entries.SOURCE_INTEGRATION_DISCOVERY
+            _LOGGER.warning(
+                "Importing Nanoleaf %s from the discovery integration", name
+            )
+            return await self.async_setup_finish()
+
+        self.context["title_placeholders"] = {"name": name}
+        return await self.async_step_link()
+
+    async def async_handle_unavailable(self):
+        """Handle unavailable device."""
+        if self.context["source"] == config_entries.SOURCE_USER:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=USER_SCHEMA,
+                errors={"base": "cannot_connect"},
+                last_step=False,
+            )
+        return self.async_abort(reason="cannot_connect")
```

Comment:
Since you are already checking for unavailable in the user step, it should be a rare enough case that they have to start over now that there aren't duplicate discoveries.